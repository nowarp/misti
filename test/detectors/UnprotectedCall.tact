contract C {
    m: map<Address, Int>;
    ms: map<Int, Cell>;

    // Entrypoints
    receive(s1: Slice) {
        let a = s1.loadAddress();
        let i = s1.loadInt(64);

        // Bad cases:
        send(SendParameters{ to: a, value: 0, bounce: false, body: emptyCell() }); // Bad
        send(SendParameters{ to: newAddress(0, 0), value: 0, bounce: false, body: s1.asCell() }); // Bad
        self.m.set(a, 42); // Bad
        self.m.del(a); // Bad
        self.ms.del(i+1); // Bad
        self.ms.set(42, s1.asCell()); // Bad

        // Guarded (good) cases:
        if (a != newAddress(0, 0)) { self.m.set(a, 42); } // OK
        self.m.set(a, 42); // OK
        if (self.checkAddr(a)) { self.m.set(a, 42); } // OK
    }

    // Utility functions
    fun checkAddr(a: Address): Bool { return a != newAddress(0, 0); }
}
