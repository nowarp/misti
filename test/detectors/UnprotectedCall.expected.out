[HIGH] UnprotectedCall: Unprotected send argument: a
test/detectors/UnprotectedCall.tact:11:34:
  10 |         // Bad cases:
> 11 |         send(SendParameters{ to: a, value: 0, bounce: false, body: emptyCell() }); // Bad
                                        ^
  12 |         send(SendParameters{ to: newAddress(0, 0), value: 0, bounce: false, body: s1.asCell() }); // Bad

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall

[HIGH] UnprotectedCall: Unprotected send argument: s1.asCell()
test/detectors/UnprotectedCall.tact:12:83:
  11 |         send(SendParameters{ to: a, value: 0, bounce: false, body: emptyCell() }); // Bad
> 12 |         send(SendParameters{ to: newAddress(0, 0), value: 0, bounce: false, body: s1.asCell() }); // Bad
                                                                                         ^
  13 |         self.m.set(a, 42); // Bad

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall

[HIGH] UnprotectedCall: Unprotected field mutation: a
test/detectors/UnprotectedCall.tact:13:20:
  12 |         send(SendParameters{ to: newAddress(0, 0), value: 0, bounce: false, body: s1.asCell() }); // Bad
> 13 |         self.m.set(a, 42); // Bad
                          ^
  14 |         self.m.del(a); // Bad

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall

[HIGH] UnprotectedCall: Unprotected field mutation: a
test/detectors/UnprotectedCall.tact:14:20:
  13 |         self.m.set(a, 42); // Bad
> 14 |         self.m.del(a); // Bad
                          ^
  15 |         self.ms.del(i+1); // Bad

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall

[HIGH] UnprotectedCall: Unprotected field mutation: i + 1
test/detectors/UnprotectedCall.tact:15:21:
  14 |         self.m.del(a); // Bad
> 15 |         self.ms.del(i+1); // Bad
                           ^
  16 |         self.ms.set(42, s1.asCell()); // Bad

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall

[HIGH] UnprotectedCall: Unprotected field mutation: s1.asCell()
test/detectors/UnprotectedCall.tact:16:25:
  15 |         self.ms.del(i+1); // Bad
> 16 |         self.ms.set(42, s1.asCell()); // Bad
                               ^
  17 | 

See: https://nowarp.io/tools/misti/docs/detectors/UnprotectedCall